src/persist/src/sqlite.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Implementation of [Consensus] backed by sqlite.
+
+use std::path::Path;
+use std::sync::Arc;
+use std::time::Instant;
+
+use async_trait::async_trait;
+use rusqlite::{named_params, params, Connection, OptionalExtension};
+use tokio::sync::Mutex;
+
+use crate::error::Error;
+use crate::storage::{Consensus, SeqNo, VersionedData};
+
+const APPLICATION_ID: i32 = 0x0678_ef32; // chosen randomly
+
+const SCHEMA: &str = "
+CREATE TABLE consensus (
+    shard text NOT NULL UNIQUE,
+    sequence_number integer NOT NULL,
+    data blob NOT NULL
+);
+";
+
+/// Implementation of [Consensus] over a sqlite database.
+#[derive(Debug)]
+pub struct SqliteConsensus {
+    conn: Arc<Mutex<Connection>>,
+    shard: String,
+}
+
+impl SqliteConsensus {
+    /// Open a sqlite-backed [Consensus] instance at `path`, for the collection
+    /// named `shard`.
+    pub fn open(path: &Path, shard: String) -> Result<Self, Error> {
+        let mut conn = Connection::open(path)?;
+        let tx = conn.transaction()?;
+        let app_id: i32 = tx.query_row("PRAGMA application_id", params![], |row| row.get(0))?;
+        if app_id == 0 {
+            tx.execute_batch(&format!(
+                "PRAGMA application_id = {APPLICATION_ID};
+                 PRAGMA user_version = 1;"
+            ))?;
+            tx.execute_batch(SCHEMA)?;
+        } else if app_id != APPLICATION_ID {
+            return Err(Error::from(format!("invalid application id: {}", app_id)));
+        }
+        tx.commit()?;
+        Ok(SqliteConsensus {
+            conn: Arc::new(Mutex::new(conn)),
+            shard,
+        })
+    }
+}
+
+#[async_trait]
+impl Consensus for SqliteConsensus {
+    async fn head(&self, _deadline: Instant) -> Result<Option<VersionedData>, Error> {
```

Comment:
is there any way for us to use the timeout? also possibly relevant for `open`

Reply:
ping on this?

Reply:
I don't think so unfortunately. I searched both the `rusqlite` and `sqlite` docs for `timeout` and `deadline` and the best I could find was https://docs.rs/rusqlite/latest/rusqlite/struct.Connection.html#method.busy_timeout which isn't really clear on what exactly it does. Like, on encountering busy, does it just wait N ms, and then retry and then if its still busy does it give up and exit, or does wake up and retry multiple times during the N ms?

Either way, it didn't seem obvious how to support our requested deadline semantics here, so opted to punt on this in this PR